### PR TITLE
feat: add boolean behavior for time objects

### LINKED ISSUE #22008: [pkg/ottl] Boolean expressions are update to be able to handle comparing `time.Time` types
In order to facilitate contexts returning time fields as `time.Time` we need to be able to compare `time.Time` values in boolean expressions.  Although not supported by the grammar as literals, OTTL should allow working with these types.

When comparing `time.Time` fields, the following rules should be followed:
- `time.Equal` should be used to check if 2 times are equal.
- `time.Before` should be used to check if 1 time is earlier than another.
- `time.After` should be used to check if 1 time is later than another.

Given 2 times, A and B:
- `A < B` is true if and only if `A.Before(B)` is true.
- `A <= B` is true if `A.Before(B)` is true OR `A.Equal(B)` is true.
- `A > B` is true if and only if `A.After(B)` is true.
- `A >= B` is true if `A.After(B)` is true OR `A.Equal(B)` is true.
- `A == B` is true if and only if `A.Equal(B)` is true.
- `A != B` is true if and only if `A.Equal(B)` is false.

If either A or B is not a time, the comparison is false.

### PR DESCRIPTION
**Pull request by fchikwekwe**
Description: Allows time comparison by enabling boolean behavior for time objects.

Link to tracking Issue: Closes #22008

Testing: Unit tests

Documentation:

### CODE REVIEW

File: pkg/ottl/compare.go [hunk 1]
```diff
@@ -135,6 +136,48 @@ func (p *Parser[K]) compareFloat64(a float64, b any, op compareOp) bool {
...
 
+func (p *Parser[K]) compareTime(a time.Time, b any, op compareOp) bool {
+	switch v := b.(type) {
+	case time.Time:
+		switch op {
+		case EQ:
+			if a.Equal(v) == true {
+				return true
+			}
+			return false
```
Comment (TylerHelmuth) [inline pkg/ottl/compare.go:147]:
```suggestion
			return a.Equal(v)
```

Let's simplify the different operation cases to return the value returned by the function.

File: pkg/ottl/compare.go [hunk 2]
```diff
@@ -135,6 +136,48 @@ func (p *Parser[K]) compareFloat64(a float64, b any, op compareOp) bool {
...
+		case LT:
+			if a.Before(v) == true {
+				return true
+			}
+			return false
+		case LTE:
+			if a.Before(v) == true || a.Equal(v) == true {
+				return true
+			}
+			return false
```
Comment (TylerHelmuth) [inline pkg/ottl/compare.go:162]: For these we can return `a.Before(v) || a.Equal(v)`

File: pkg/ottl/boolean_value_test.go [hunk 3]
```diff
@@ -106,6 +119,13 @@ func Test_newComparisonEvaluator(t *testing.T) {
 		{name: "[]byte('a') < []byte('b')", l: []byte("a"), r: []byte("b"), op: "<", want: true},
 		{name: "nil == nil", op: "==", want: true},
 		{name: "nil == []byte(nil)", r: []byte(nil), op: "==", want: true},
+		{name: "compare equal times", l: "time1", r: "time2", op: "==", want: true, item: map[string]time.Time{"time1": JanFirst2023, "time2": JanFirst2023}},
+		{name: "compare unequal times", l: "time1", r: "time2", op: "==", want: false, item: map[string]time.Time{"time1": JanFirst2023, "time2": time.Date(2023, 1, 2, 0, 0, 0, 0, time.Local)}},
+		{name: "compare for inequality w/ times", l: "time1", r: "time2", op: "!=", want: true, item: map[string]time.Time{"time1": JanFirst2023, "time2": time.Date(2002, 11, 2, 01, 01, 01, 01, time.Local)}},
```
Comment (TylerHelmuth) [inline pkg/ottl/boolean_value_test.go:124]:
```suggestion
		{name: "compare for not equal times", l: "time1", r: "time2", op: "!=", want: true, item: map[string]time.Time{"time1": JanFirst2023, "time2": time.Date(2002, 11, 2, 01, 01, 01, 01, time.Local)}},
```

File: pkg/ottl/boolean_value_test.go [hunk 4]
```diff
@@ -106,6 +119,13 @@ func Test_newComparisonEvaluator(t *testing.T) {
...
+		{name: "compare less than w/ times", l: "time1", r: "time2", op: "<", want: true, item: map[string]time.Time{"time1": JanFirst2023, "time2": time.Date(2023, 5, 2, 01, 01, 01, 01, time.Local)}},
+		{name: "compare less than equal to w/ times", l: "time1", r: "time2", op: "<=", want: true, item: map[string]time.Time{"time1": time.Date(2002, 5, 2, 01, 01, 01, 01, time.Local), "time2": time.Date(2003, 5, 2, 01, 01, 01, 01, time.Local)}},
+		{name: "compare greater than equal to w/ times", l: "time1", r: "time2", op: ">=", want: false, item: map[string]time.Time{"time1": time.Date(2002, 5, 2, 01, 01, 01, 01, time.Local), "time2": time.Date(2003, 5, 2, 01, 01, 01, 01, time.Local)}},
+		{name: "compare greater than w/ times", l: "time1", r: "time2", op: ">", want: true, item: map[string]time.Time{"time1": time.Date(2022, 5, 2, 01, 01, 01, 01, time.Local), "time2": time.Date(2003, 5, 2, 01, 01, 01, 01, time.Local)}},
```
Comment (TylerHelmuth) [inline pkg/ottl/boolean_value_test.go:128]: Can you add more tests cases such that each operator has 2 tests, 1 where we expect `true` and one where we expect `false`?

File: pkg/ottl/boolean_value_test.go (hunk 3 above)
Comment (fchikwekwe) [reply pkg/ottl/boolean_value_test.go]: Updated.

File: pkg/ottl/boolean_value_test.go (hunk 4 above)
Comment (fchikwekwe) [reply pkg/ottl/boolean_value_test.go:132]: I added more tests where want is true and false for each operator.

File: pkg/ottl/compare.go (hunk 2 above)
Comment (fchikwekwe) [reply pkg/ottl/compare.go]: Done!

File: pkg/ottl/compare.go [hunk 5]
```diff
@@ -135,6 +136,30 @@ func (p *Parser[K]) compareFloat64(a float64, b any, op compareOp) bool {
...
+	case time.Time:
+		switch op {
+		case EQ:
+			return a.Equal(v)
+		case NE:
+			return !a.Equal(v)
+		case LT:
+			return a.Before(v)
+		case LTE:
+			return a.Before(v) == true || a.Equal(v) == true
```
Comment (TylerHelmuth) [inline pkg/ottl/compare.go:150]:
```suggestion
			return a.Before(v) || a.Equal(v)
```

File: pkg/ottl/compare.go [hunk 6]
```diff
@@ -135,6 +136,30 @@ func (p *Parser[K]) compareFloat64(a float64, b any, op compareOp) bool {
...
+		case GTE:
+			return a.After(v) == true || a.Equal(v) == true
```
Comment (TylerHelmuth) [inline pkg/ottl/compare.go:152]:
```suggestion
			return a.After(v) || a.Equal(v)
```

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**